### PR TITLE
feat: drop node4 support

### PR DESCRIPTION
**Pull request by aaarichter**
Changes:
- [x] drop node4 support
- [x] use Map API instead of object hash maps
- [x] remove lodash dependency

### CODE REVIEW

File: lib/cached.js [hunk 1]
```diff
@@ -68,17 +71,17 @@ cached.createCache = function createCache(options) {
...
 
 cached.dropNamedCache = function dropNamedCache(name) {
-  delete namedCaches[name];
+  namedCaches.delete(name);
   return cached;
 };
 
 cached.knownCaches = function knownCaches() {
-  return Object.keys(namedCaches);
+  return [...namedCaches.keys()];
```
Comment (aaarichter) [inline lib/cached.js]: Q: should we return a MapIterator instead ?

File: lib/get-or-else.js [hunk 2]
```diff
@@ -77,10 +76,12 @@ function getOrElse(cache, key, val, opts) {
 
   function verifyFreshness(wrappedValue) {
     const hit = !!wrappedValue;
-    const expired = util.isExpired(wrappedValue && wrappedValue.b);
     let loadingNewValue = cache.staleOrPending[key] !== undefined;
 
-    if ((!hit || expired) && !loadingNewValue) {
+    if (
```
Comment (jkrems) [inline lib/get-or-else.js]: Any reason why we went back to a huge if condition here?

File: lib/cached.js (hunk 1 above)
Comment (jkrems) [reply lib/cached.js]: We wanted to keep the interface slim. Exposing a map iterator would lock us into using a `Map` internally in future versions as well. I think using an array here is great (a `Set` would also work but might be unnecessary overhead).

File: lib/backends/memcached.js [hunk 3]
```diff
@@ -63,11 +62,11 @@ MemcachedBackend.prototype.get = function get(key) {
 };
 
 MemcachedBackend.prototype.set = function set(key, value, options) {
-  return this._clientSet(key, value, options.expire).then(_.constant(value));
```
Comment (jkrems) [inline lib/backends/memcached.js]: nit: here and below, we're now creating new functions for every call. Also: `_.constant` is *not* an identity function. This didn't used to rely on `_clientSet` returning the value but was always returning `value` from the function arguments.

File: lib/get-or-else.js (hunk 2 above)
Comment (aaarichter) [reply lib/get-or-else.js]: so the reason for the change is the likelihood of each part on the condition as well as calling functions / getting values

- cache.staleOrPending is set -> skip the more complex condition `&& (hit || expired)`
- hit: if no hit then `expired` does not need to be calculated
- `util.isExpired(wrappedValue && wrappedValue.b)` don't check the value if any of the other cheaper checks pass

File: lib/backends/memcached.js (hunk 3 above)
Comment (aaarichter) [reply lib/backends/memcached.js]: good catch - it should be `() => value`

File: lib/backends/memcached.js [hunk 4]
```diff
@@ -33,9 +33,10 @@
 'use strict';
 
 const promisify = require('bluebird').promisify;
-const _ = require('lodash');
 const Memcached = require('memcached');
 
+const noop = () => undefined;
```
Comment (jkrems) [inline lib/backends/memcached.js]: Side note: Using `function` would have been shorter and wouldn't have closed over the weird module-scope `this`.

```js
function noop() {}
const noop = () => undefined;
```

File: lib/cached.js [hunk 5]
```diff
@@ -33,33 +33,36 @@
...
 
 const Cache = require('./cache');
 
-let namedCaches = Object.create(null);
+const namedCaches = new Map();
 
-const getName = util.deprecate(function getName(name) {
-  return name || 'default';
-}, 'Unnamed caches and caches with non-string names are deprecated.');
+const getName = util.deprecate(
```
Comment (jkrems) [inline lib/cached.js]: nit: This turns a previously named function into something that won't have a meaningful `.name` anymore which is a little worse in terms of debugging.

File: test/timeout.test.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import assert from 'assertive';
 import Bluebird from 'bluebird';
```
Comment (jkrems) [inline test/timeout.test.js]: Huh, we should undo this eventually. Completely forgot that we were using babel script modules in this project. Good reminder. :)

**Pull request merged**